src/test/func/malloc/malloc.cc

```diff
@@ -0,0 +1,127 @@
+#include <cassert>
+#include <cerrno>
+#include <ctime>
+#include <malloc.h>
+#include <snmalloc.h>
+#include <test/xoroshiro.h>
+
+using namespace snmalloc;
+
+constexpr size_t min_pow_2 = bits::ctz_const(sizeof(size_t));
+constexpr size_t max_pow_2 = bits::ctz_const(SUPERSLAB_SIZE);
+
+size_t rand_pow_2(xoroshiro::p128r64& r, size_t min_pow = min_pow_2)
+{
+  const size_t max_pow = max_pow_2 - min_pow;
+  auto p = (r.next() % (max_pow + 1)) + min_pow;
+  return (size_t)1 << p;
+}
+
+size_t rand_non_pow_2(xoroshiro::p128r64& r, size_t min = 0)
+{
+  const size_t max = (1 << max_pow_2) - min;
+  auto n = (r.next() % (max + 1)) + min;
+  while (__builtin_popcount(n) < 2)
+    n = r.next();
+
+  return n;
+}
+
+void check_result(size_t size, void* p, int err, bool null)
+{
+  fprintf(stderr, "%p, errno: %s\n", p, strerror(errno)); // TODO
+  assert(errno == err);
+  UNUSED(err);
+  if (null)
+  {
+    assert(p == nullptr);
+  }
+  else
+  {
+    assert(malloc_usable_size(p) >= size);
+    UNUSED(size);
+    free(p);
+  }
+}
+
+void test_calloc(size_t nmemb, size_t size, int err, bool null)
+{
+  fprintf(stderr, "calloc(%d, %d)\n", (int)nmemb, (int)size);
+  errno = 0;
+  void* p = calloc(nmemb, size);
+
+  if ((p != nullptr) && (errno == 0))
+    for (size_t i = 0; i < (size * nmemb); i++)
+      assert(((uint8_t*)p)[i] == 0);
+
+  check_result(nmemb * size, p, err, null);
+}
+
+void test_realloc(void* p, size_t size, int err, bool null)
+{
+  fprintf(stderr, "realloc(%p, %d)\n", p, (int)size);
+  errno = 0;
+  p = realloc(p, size);
+  check_result(size, p, err, null);
+}
+
+void test_posix_memalign(size_t size, size_t align, int err, bool null)
+{
+  fprintf(stderr, "posix_memalign(&p, %d, %d)\n", (int)align, (int)size);
+  void* p = nullptr;
+  errno = posix_memalign(&p, align, size);
+  check_result(size, p, err, null);
+  if (!null)
+    assert(((uintptr_t)p % align) == 0);
+}
+
+void test_memalign(size_t size, size_t align, int err, bool null)
+{
+  fprintf(stderr, "memalign(%d, %d)\n", (int)align, (int)size);
+  errno = 0;
+  void* p = memalign(align, size);
+  check_result(size, p, err, null);
+  if (!null)
+    assert(((uintptr_t)p % align) == 0);
+}
+
+int main(int argc, char** argv)
+{
+  UNUSED(argc);
+  UNUSED(argv);
+
+  constexpr int SUCCESS = 0;
+
+  auto r = xoroshiro::p128r64((size_t)time(nullptr));
```

Comment:
I don't like the test being unpredictable.  This may or may not fail in CI.  This is bad for reproducibility.

I would sooner you take a seed from the command line, and run multiple instances of the test with different seeds.

If you really think having some "real" randomness is necessary, please print the seed, and allow it to be taken from the command line, so we can reproduce failures.

Reply:
The randomness wasn't necessary. The tests now run through all size classes and power of 2 alignments that may have been chosen randomly